lib/srv/app/server.go

```diff
@@ -802,8 +803,36 @@ func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		// Covert trace error type to HTTP and write response, make sure we close the
 		// connection afterwards so that the monitor is recreated if needed.
 		code := trace.ErrorToCode(err)
+
+		// Return a custom message for device trust errors.
+		var text string
+		if errors.Is(err, services.ErrTrustedDeviceRequired) {
+			// Try to guess the app from the host for a better error messages.
+			hostOnly, _, _ := net.SplitHostPort(r.Host)
+			appName := "yourapp"
+			for _, app := range s.getApps() {
+				if app.GetPublicAddr() == hostOnly || app.GetPublicAddr() == r.Host {
+					appName = app.GetName()
+					break
+				}
+			}
+
+			text = fmt.Sprintf(`Access to this app requires a trusted device.
+
+Try running:
+
+tsh proxy app %s -p 8888
```

Comment:
> So if the user tries to App Access through Teleport Web App, it will not work and receive this?

Correct, assuming they have a role that says device_trust_mode=required.

> How about using curl with the certs from tsh apps login?

Yes, that works.

> All app access using tsh local proxies should work out of the box (e.g tsh aws)?

Correct, also works. `tsh` is capable of device authentication, so the certs we get from it are good. The Web UI, conversely, cannot do device authentication and thus fails when the check is required.

Do note that the user [must enroll their device][1] for it to work. There are features that it make it rather easy, [like auto-enrollment][2], but ultimately depends on cluster configs.


[1]: https://goteleport.com/docs/access-controls/device-trust/device-management/#create-a-device-enrollment-token
[2]: https://goteleport.com/docs/access-controls/device-trust/device-management/#auto-enrollment

Reply:
As long as the AWS console isn't targeted by a device_trust_mode=required role, then it should be alright.

Let me know if you'd like to see any follow ups on this, maybe a docs mention?

Reply:
Yes, they would get the new message as of this commit. Chatting out-of-band about this.

Reply:
I've simplified the message with the intention of pushing more specific instructions into the docs link. This makes it a bit less useful by itself, but now it doesn't run the risk of being incorrect (like for AWS Web Console, future Azure/GCP consoles, `tsh` using a non-authorized device, etc).